src/Domain/UseCase/CreateValetToken/CreateValetToken.ts

```diff
@@ -0,0 +1,56 @@
+import { inject, injectable } from 'inversify'
+import { isOperationPermitted } from '../../Operation/isOperationPermitted'
+import { ValetToken } from '../../ValetToken/ValetToken'
+import { CreateValetKeyDto } from './CreateValetTokenDto'
+import { CreateValetKeyResponse } from './CreateValetTokenResponse'
+import { UseCaseInterface } from '../UseCaseInterface'
+import { generateValetToken } from '../../ValetToken/generateValetToken'
+import TYPES from '../../../Bootstrap/Types'
+import { JwtSecret, ValetTokenSecret } from '../../ValetToken/ValetTokenGenerator'
+import { CrypterInterface } from '../../Encryption/CrypterInterface'
+import { createValetPayload } from '../../ValetToken/createValetPayload'
+
+@injectable()
+export class CreateValetToken implements UseCaseInterface {
+  constructor(
+    @inject(TYPES.JWT_SECRET) private jwtSecret: JwtSecret,
+    @inject(TYPES.VALET_TOKEN_SECRET) private valetTokenSecret: ValetTokenSecret,
+    @inject(TYPES.Crypter) private crypter: CrypterInterface,
+  ) {}
+
+  /**
+   * Given a request from the API Gateway containing auth data for a user and the requested file operations generates a valet key which can be used to perform these operations directly via the Files Service.
+   */
+  async execute(dto: CreateValetKeyDto): Promise<CreateValetKeyResponse> {
+    const { user, operation, resources, validityPeriod } = dto
+    const { permissions, uuid } = user
+
+    if (!isOperationPermitted({
+      operation,
+      permissions,
+      resources,
+    })) return {
+      success: false,
+      error: {
+        message: `User ${uuid} is not authorized to perform the requested operation.`,
+        forbiddenOperation: operation,
+      },
+    }
+
+    const payload = createValetPayload({
+      uuid, 
+      permittedOperation: operation,
+      permittedResources: resources,
+      validityPeriod,
+    })
+
+    const valetToken: ValetToken = await generateValetToken({
+      payload,
+      jwtSecret: this.jwtSecret,
+      valetTokenSecret: this.valetTokenSecret,
+      crypter: this.crypter,
```

Comment:
if you've got `this.valetTokenGenerator` you won't need to pass `this.crypter` there. Now you have what's called a hidden dependency whereas if you turn that function (`generateValetToken`) into a generator class you will explicitly state the dependency there which is `CrypterInterface`

Reply:
The `crypter` dependency is explicit in the constructor of the use case and in the function signature. It is injected into the class and passed to the function from there. Both are testable independently.

If I turn the function into a class I will only move the dependency from one parameter list to another (function to constructor). Yes, thanks to that I can then add an annotation for the framework to inject it for me, but this is not necessary in this case, so there is no point in paying the complexity price for that.

The essential point is the same as in https://github.com/standardnotes/files/pull/3#discussion_r598958158